### PR TITLE
i#3544 riscv64, part 2: add instr_info_t generator

### PR DESCRIPTION
**Pull request by semihalf-kardach-stanislaw**
Add a script for parsing Instruction Set Listing files. The goal is to generate:

1. A list of supported instruction opcodes (opcode_api.h).
2. A full set of instruction creation macros (instr_create_api.h).
3. An instr_info_t table and a lookup trie for all known instructions.

The first two files are generated based on template files so that they may include non-generated data as well. The template files will be added in following commits.
The last file is generated from scratch.

Issue #3544

Signed-off-by: Stanislaw Kardach <kda@semihalf.com>

### CODE REVIEW

**Comment by derekbruening** [comment]
@bete0 had volunteered to review this one

File: core/ir/riscv64/codec.py [hunk 1]
```diff
@@ -0,0 +1,1064 @@
...
+            # Immediate argument will handle the base+disp
+            inst.flds.pop(1)
+            dbg(f'    -> {" " * len(inst.name)} {[f.name for f in inst.flds]}')
+        elif (Field.CB_IMM in inst.flds):
+            # Compare-and-branch instructions need their branch operand moved
+            # to the 1st source operand slot as required by instr_set_target().
+            dbg(f'fixup: {inst.name} {[f.name for f in inst.flds]}')
+            cb_imm = inst.flds.pop(0)
+            inst.flds.append(cb_imm)
+            dbg(f'    -> {" " * len(inst.name)} {[f.name for f in inst.flds]}')
```
Comment (bete0) [inline core/ir/riscv64/codec.py:613]: Do we need an else clause to raise an exception or sth if  neither or these conditions are hit? Same below.

**Review by bete0 — CHANGES_REQUESTED** [review]

File: core/ir/riscv64/codec.py [hunk 2]
```diff
@@ -0,0 +1,1064 @@
...
+        - LUI's U_IMM operand is OPSZ_20b and not OPSZ_4 because assembly takes
+          the upper 20 bits of the created immediate.
+        - C.LD's CLD_IMM is OPSZ_1 and not OPSZ_5b because assembly takes the
+          full 1-byte offset value, not the top 5 bits of it.
+        - BLT's B_IMM is OPSZ_2 because the target address must be 2-byte
+          aligned.
+        '''
+        if (type(self.opsz_def) is str):
+            return self.opsz_def
+        else:
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py:519]: Nope. I'll remove it.

File: core/ir/riscv64/codec.py (hunk 1 above)
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py:613]: No. The "else" case means that a given instruction doesn't need a fixup.

File: core/ir/riscv64/codec.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done

File: core/ir/riscv64/codec.py [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+        Immediate operand sizes depend on the instruction to match the assembly
+        if possible. I.e.:
+        - LUI's U_IMM operand is OPSZ_20b and not OPSZ_4 because assembly takes
+          the upper 20 bits of the created immediate.
+        - C.LD's CLD_IMM is OPSZ_1 and not OPSZ_5b because assembly takes the
+          full 1-byte offset value, not the top 5 bits of it.
+        - BLT's B_IMM is OPSZ_2 because the target address must be 2-byte
+          aligned.
+        '''
+        if (type(self.opsz_def) is str):
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done

**Comment by semihalf-kardach-stanislaw** [comment]
@bete0 could you check this PR again? I believe I've fixed all the issues you've mentioned.

**Comment by derekbruening** [comment]
@bete0 is out of town all week.  Maybe someone else can look at it..

**Comment by derekbruening** [comment]
> @bete0 is out of town all week. Maybe someone else can look at it..

I will take a look.

File: core/ir/riscv64/codec.py [hunk 4]
```diff
@@ -0,0 +1,1063 @@
...
+        Generate the opcode_api.h header file.
+
+        Returns true if the output file was written successfully.
+        '''
+        tmpl_fld = '@OPCODES@'
+        if len(self.instructions) == 0:
+            return False
+        found_template_fld = False
+        idx = -1
+        # Replace tmpl_fld in the template_file and save as out_file
```
Comment (derekbruening) [inline core/ir/riscv64/codec.py:764]: nit: Our style guide says to prefer including end punctuation with complete sentences in comments -- here and elsewhere

File: core/ir/riscv64/codec.py [hunk 5]
```diff
@@ -0,0 +1,1063 @@
...
+                            args += ', '.join([f.formatted_name()
+                                              for f in flds])
+                            arg_comments += '\n'
+                            arg_comments += '\n'.join(
+                                [f' * \param {f.formatted_name():6}  {f.arg_cmt}' for f in flds])
+                        nd = len([f for f in flds if f.is_dest])
+                        ns = len(flds) - nd
+                        lines.append(
+                            f'''/**
+ * Creates a {i.name} instruction.
```
Comment (derekbruening) [inline core/ir/riscv64/codec.py:829]: s/a/a(n)/ so nobody complains about the wrong article before a vowel?  Or I guess you could actually check whether i.name starts with a vowel.

File: core/ir/riscv64/codec.py [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+                continue
+            bucket_size = mask + 1
+            dbg(f'mask:        {mask:032b}')
+            dbg(f'shift:       {shift}')
+            dbg(f'bucket size: {bucket_size}')
+            buckets: List[List[Instruction]] = [[]
+                                                for i in range(0, bucket_size)]
+            # Each trie bucket contains a list of instructions sharing
+            # (inst.match >> shift) & mask, where mask is a continuous set of 1
+            # bits.
```
Comment (derekbruening) [inline core/ir/riscv64/codec.py:881]: nit: s/continuous/contiguous/

File: core/ir/riscv64/codec.py [hunk 7]
```diff
@@ -0,0 +1,1063 @@
...
+            # There is a special case when trie creation algorithm detects
+            # aliased instructions (that is having different inst.mask but the
+            # ame inst.match - i.e. ori and prefetch.[irw]). In that case bucket
```
Comment (derekbruening) [inline core/ir/riscv64/codec.py:884]: typo: s/ame/same/

**Review by derekbruening — APPROVED** [review]

**Comment by derekbruening** [comment]
Hmm, didn't realize my approval would not remove the "changes requested" from @bete0 .  I would say go ahead and get it all ready to commit and we can override the block at that point.

**Review by bete0 — APPROVED** [review]

File: core/ir/riscv64/codec.py [hunk 8]
```diff
@@ -0,0 +1,1063 @@
...
+            'OPSZ_2b',
+            '',
+            'The acquire-release constraint field (inst[26:25]).'
+            )
+    CSR = (13,
+           'csr',
+           False,
+           'OPSZ_PTR',
+           '',
+           'The Configuration/Status Register id (inst[31:20]).'
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py:213]: To emphasize this is a `CSR` but given the parameter name I don't think it's necessary.

File: core/ir/riscv64/codec.py [hunk 9]
```diff
@@ -0,0 +1,1063 @@
...
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# This script generates encoder/decoder logic for RISC-V instructions described
+# in core/arch/riscv64/isl/*.txt files. This includes:
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done, thanks for spotting!

File: core/ir/riscv64/codec.py [hunk 10]
```diff
@@ -0,0 +1,1063 @@
...
+# - instr_create_api.h - Instruction generation macros.
+# - FIXME i#3544: add a list of generated files here.
+#
+# FIXME i#3544: To be done:
+# - Generate instruction encoding (at format + fields level?).
+# - Gather statistics on instruction usage frequency. If there is a small set
+#   of instructions which are used most frequently, then having an
+#   instruction-specific [de]ncoder will utilize branch predictor better and not
+#   cause dramatically higher instruction cache polution than a per-format
+#   decoding functions with per-field branches.
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py:45]: Done.

File: core/ir/riscv64/codec.py [hunk 11]
```diff
@@ -0,0 +1,1063 @@
...
+log_handler = StreamHandler()
+log_handler.setFormatter(Formatter('%(levelname)s: %(message)s'))
+logger.addHandler(log_handler)
+
+
+def dbg(msg, *args, **kwargs):
+    logger.debug(msg, *args, **kwargs)
+
+
+def wrn(msg, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py [hunk 12]
```diff
@@ -0,0 +1,1063 @@
...
+    logger.warning(msg, *args, **kwargs)
+
+
+def nfo(msg, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py [hunk 13]
```diff
@@ -0,0 +1,1063 @@
...
+def err(msg, *args, **kwargs):
+    logger.error(msg, *args, **kwargs)
+
+
+env_debug = getenv('DEBUG') or '0'
+if env_debug != '0':
+    logger.setLevel(DEBUG)
+
+
+def maxu(bits: int) -> int:
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py [hunk 14]
```diff
@@ -0,0 +1,1063 @@
...
+             'OPSZ_5b',
+             '',
+             'The `shamt` field (bit range is determined by XLEN).'
+             )
+    SHAMT5 = (16,
+              'shamt',
+              False,
+              'OPSZ_6b',
+              '',
+              'The `shamt` field that uses 5-bits.'
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py (hunk 4 above)
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done. I think I've caught all.

File: core/ir/riscv64/codec.py (hunk 6 above)
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py:884]: Done.

File: core/ir/riscv64/codec.py (hunk 7 above)
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py (hunk 5 above)
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: The problem here is that we should use "an" before a "vowel sound", not a vowel per-se (i.e. "a union") so I went with `a(n)`.

File: core/ir/riscv64/codec.py [hunk 15]
```diff
@@ -0,0 +1,1063 @@
...
+env_debug = getenv('DEBUG') or '0'
+if env_debug != '0':
+    logger.setLevel(DEBUG)
+
+
+def maxu(bits: int) -> int:
+    return (1 << bits) - 1
+
+
+def ctz(n: int) -> int:
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

File: core/ir/riscv64/codec.py [hunk 16]
```diff
@@ -0,0 +1,1063 @@
...
+    sn = bin(n)
+    return len(sn) - len(sn.rstrip('0'))
+
+
+def cto(n: int) -> int:
```
[the remark this answers is not in the document]
Comment (semihalf-kardach-stanislaw) [reply core/ir/riscv64/codec.py]: Done.

**Pull request merged**